Including common Non SFA functions

situx opened the pull request:
Closes #257 
Closes #341
Closes #400
- aggCollect function
- DWithin function
- metricDWithin function
- isCollection function
- simplifyDP function

nicholascar reviewed (changes requested):
- [x] Why are acronyms used here like `DWithin` when previous functions all use full wording such as `hasMetricParameterLength`? Prefer consistencey across all functions unless a justification is provided
- [ ] Reference for the Douglas-Peucker algorithm needed

situx commented:
Valid points.
Should we name it geo:DistanceWithin then?

mperry455 commented on spec/sections/11-geometry_extension.adoc:
````diff
@@ -1952,6 +1993,14 @@ geof:aggCentroid (geom: ogc:geomLiteral): ogc:geomLiteral
+geof:aggCollect (geom: ogc:geomLiteral): ogc:geomLiteral
+```
+
+The function http://www.opengis.net/def/function/geosparql/aggCollect[`geof:aggCollect`] calculates creates a GeometryCollection literal out of a set of geometries.
````
Typo: "calculates creates"

VladimirAlexiev replied:
does `geom: ogc:geomLiteral` indicate that it's an array?

situx replied:
It is a set of ogc:geomLiterals in pratctice likely bound to a SPARQL variable like in other functions too.
The content of the ogc:geomLiterals could be arrays, e.g. a WKT GeometryCollection could be considered as an array.
Does that answer your question or would you have a different idea how to mark an array/set?

VladimirAlexiev replied:
@situx I saw your answer in another issue that one geomLiteral can carry a collection. 
But I think this is different: aggregates work on a resultset of multiple geomLiterals, so the function signature should indicate an array. Eg use `geomLiteral*`

VladimirAlexiev replied:
My comment is answered. But @mperry455 's comment not answered yet

mperry455 reviewed (approved):
Approved pending suggested changes by @nicholascar

mperry455 commented:
I feel like geof:withinDistance reads a little better than geof:distanceWithin.

VladimirAlexiev commented:
Rename `simplifyDP` to `simplify`. When another (less popular algorithm) needs to be added, it can get a longer name `simplifyFoo`.

Simplify! :-)

jabhay commented:
Discussed simplifyDP function. Given it's the only simplification function being considered currently, we arrived at the decision to call it just `simplify`.

situx commented:
I applied all changes requested. With that we can re-review this pull request

VladimirAlexiev reviewed (changes requested):
geomLiteral array

jabhay commented:
Would be good to get a few eyes on the following options:

1. Recommended above: geomLiteral*
2. Other option in line with other programming languages: geomLiteral[]

Can people add their thoughts to this before next meeting?

jabhay commented:
1. Recommended above: geomLiteral*

jabhay commented:
3. Leave as is, and explain it

jabhay commented:
Voting result: annotate as ogc:geomLiteral[] and add explanatory text to describe it.

situx commented:
I added the array notation to all spatial aggregate functions, added a paragraph to explain the syntax and corrected some mistakes. The PR is therefore again ready for review.

mperry455 commented on spec/sections/11-geometry_extension.adoc:
```diff
@@ -1203,6 +1203,8 @@ Implementations shall support the functions
 <<Function: geof:metricBuffer, `geof:metricBuffer`>>, 
 <<Function: geof:buffer, `geof:buffer`>>, 
+<<Function: geof:metricWithinDistance, `geof:metricWithinDistance`>>, 
+<<Function: geof:WithinDistance, `geof:WithinDistance`>>, 
```
Lowercase 'w'? geof:withinDistance instead of geof:WithinDistance

mperry455 commented on spec/sections/11-geometry_extension.adoc:
```diff
@@ -1442,6 +1446,24 @@ geof:coordinateDimension (geom: ogc:geomLiteral): xsd:integer
+
+Returns true if `geom2` is within a given distance of `geom1`. Calculations are according to the units in the spatial reference system of `geom1`. The triple store implementation needs to take care of a conversion between meter and the unit of the spatial reference system.
+
+==== Function: geof:sfWithinDistance
```
geof:withinDistance?

VladimirAlexiev commented on spec/sections/07-conventions.adoc:
```diff
@@ -57,6 +57,12 @@ The following IRI namespace prefixes are used throughout this document:
 
+=== Data Types for Spatial Aggregate Functions
+
+In this specification we use the placeholder URI ogc:geomLiteral to describe any geometry literal which is defined in this specification.
```
which -> that

VladimirAlexiev commented on spec/sections/07-conventions.adoc:
```diff
@@ -57,6 +57,12 @@ The following IRI namespace prefixes are used throughout this document:
+=== Data Types for Spatial Aggregate Functions
+
+In this specification we use the placeholder URI ogc:geomLiteral to describe any geometry literal which is defined in this specification.
+To express a list of such literals, for example as input parameters for spatial aggregate functions, we use the notation ogc:geomLiteral[].
```
don't you need to put ogc:geomLiteral[] in some code markers? (I'm not familiar with adoc)

situx replied:
I do not think so. However, should we find it disturbing we can still correct the styling later on

VladimirAlexiev commented on spec/sections/07-conventions.adoc:
```diff
@@ -57,6 +57,12 @@ The following IRI namespace prefixes are used throughout this document:
+
+In this specification we use the placeholder URI ogc:geomLiteral to describe any geometry literal which is defined in this specification.
+To express a list of such literals, for example as input parameters for spatial aggregate functions, we use the notation ogc:geomLiteral[].
+We do not define the order of ogc:geomLiteral[]. The order is to be determined by implementers.
```
I don't understand that. A "list" is ordered, so how do you mean implementers can change it?

situx replied:
A list is ordered. But how it is ordered, e.g., by URI alphabetically or not at all, should be up to the implementers.

VladimirAlexiev commented on spec/sections/11-geometry_extension.adoc:
````diff
@@ -1442,6 +1446,24 @@ geof:coordinateDimension (geom: ogc:geomLiteral): xsd:integer
+                 distance: xsd:double): xsd:boolean
+```
+
+Returns true if `geom2` is within a given distance of `geom1`. Calculations are according to the units in the spatial reference system of `geom1`. The triple store implementation needs to take care of a conversion between meter and the unit of the spatial reference system.
````
"given distance of `geom1`" -> add "in meters"

nicholascar commented:
> Per SWG discussion on 10/16/2024: Regarding ordering of ogc:geometryLiteral[] input to aggregates to make result CRS deterministic, it may be hard to enforce an ordering on implementers. Here are some possible options:
> 
>     1. Leave it up to implementers to determine the output CRS
> 
>     2. Add a new target_crs argument to aggregate functions
> 
>     3. Document that users can transform the output of the aggregate to their desired CRS
> 
>     4. Specify an ordering criteria
> 
> 
> Please comment on this thread with opinions on the options.

I vote for:

3. Document that users can transform the output of the aggregate to their desired CRS

(if the implementers support this functionality - probably will)

ultimately

1. Leave it up to implementers to determine the output CRS

jabhay commented:
Vote regarding ordering of ogc:geometryLiteral[] input to aggregates to make result CRS deterministic. Thumbs up is in favour, Thumbs down is have objections.

jabhay commented:
Option 1: Leave it up to implementers to determine the output CRS

jabhay commented:
Option 2: Add a new target_crs argument to aggregate functions

jabhay commented:
Option 3: Document that users can transform the output of the aggregate to their desired CRS

jabhay commented:
Option 4: Specify an ordering criteria

situx commented:
I added a comment on parameters of spatial aggregate functions, so this pull request is now again ready for review

mperry455 reviewed (approved):
Looks good.

jabhay commented:
@VladimirAlexiev : could you please review and either approve or comment on this PR. We'd like to ensure it addresses your original concerns.
